src/language-js/postprocess.js

```diff
@@ -188,4 +191,21 @@ function includeShebang(ast, options) {
   }
 }
 
+// This is a workaround to transform `ChainExpression` from `typescript-estree` into
+// `babel` shape AST, we should do the opposite, since `ChainExpression` is the
+// standard `estree` AST for `optional chaining`
+// https://github.com/estree/estree/blob/master/es2020.md
+function transformChainExpression(node) {
+  if (node.type === "CallExpression") {
+    node.type = "OptionalCallExpression";
+    node.callee = transformChainExpression(node.callee);
+  } else if (node.type === "MemberExpression") {
+    node.type = "OptionalMemberExpression";
+    node.object = transformChainExpression(node.object);
+  } else if (node.type === "TSNonNullExpression") {
+    node.expression = transformChainExpression(node.expression);
+  }
```

Comment:
I'm not familiar with typescript, but is it possible to have other types inside `ChainExpression`?

Reply:
I'm wondering is this a mistake in parser? `ChainExpression` should be the root of `optional chaining`.

Shouldn't it be `TSNonNullExpression > ChainExpression`?

Reply:
@bradzacher 

What should this `c?.d!` be? The estree says 

> The `ChainExpression` node contains one or more `ChainElement` nodes that are `optional:true`.

https://github.com/estree/estree/blob/master/es2020.md#chainexpression

But in this case `TSNonNullExpression` is not `ChainElement`.

I'm not familiar with typescript, is this `c?.d!` mean `d` is non-null, and `(c?.d)!` mean `c.d` is non-null?

If it's correct, shouldn't `TSNonNullExpression` be the `MemberExpression.property`?

Reply:
I mirrored babel's (v7.11.3) implementation here.
This is the same AST they produce.

input code:
```ts
(a?.b!).c;
(a?.b)!.c;
```

<details>
<summary>`@babel/parser` @ `7.11.3`</summary>

```json
{
  "type": "Program",
  "start": 0,
  "end": 21,
  "loc": {
    "start": {
      "line": 1,
      "column": 0
    },
    "end": {
      "line": 2,
      "column": 10
    }
  },
  "sourceType": "module",
  "interpreter": null,
  "body": [
    {
      "type": "ExpressionStatement",
      "start": 0,
      "end": 10,
      "loc": {
        "start": {
          "line": 1,
          "column": 0
        },
        "end": {
          "line": 1,
          "column": 10
        }
      },
      "expression": {
        "type": "MemberExpression",
        "start": 0,
        "end": 9,
        "loc": {
          "start": {
            "line": 1,
            "column": 0
          },
          "end": {
            "line": 1,
            "column": 9
          }
        },
        "object": {
          "type": "ChainExpression",
          "start": 1,
          "end": 6,
          "loc": {
            "start": {
              "line": 1,
              "column": 1
            },
            "end": {
              "line": 1,
              "column": 6
            }
          },
          "expression": {
            "type": "TSNonNullExpression",
            "start": 1,
            "end": 6,
            "loc": {
              "start": {
                "line": 1,
                "column": 1
              },
              "end": {
                "line": 1,
                "column": 6
              }
            },
            "expression": {
              "type": "MemberExpression",
              "start": 1,
              "end": 5,
              "loc": {
                "start": {
                  "line": 1,
                  "column": 1
                },
                "end": {
                  "line": 1,
                  "column": 5
                }
              },
              "object": {
                "type": "Identifier",
                "start": 1,
                "end": 2,
                "loc": {
                  "start": {
                    "line": 1,
                    "column": 1
                  },
                  "end": {
                    "line": 1,
                    "column": 2
                  },
                  "identifierName": "a"
                },
                "name": "a"
              },
              "computed": false,
              "property": {
                "type": "Identifier",
                "start": 4,
                "end": 5,
                "loc": {
                  "start": {
                    "line": 1,
                    "column": 4
                  },
                  "end": {
                    "line": 1,
                    "column": 5
                  },
                  "identifierName": "b"
                },
                "name": "b"
              },
              "optional": true
            }
          },
          "extra": {
            "parenthesized": true,
            "parenStart": 0
          }
        },
        "computed": false,
        "property": {
          "type": "Identifier",
          "start": 8,
          "end": 9,
          "loc": {
            "start": {
              "line": 1,
              "column": 8
            },
            "end": {
              "line": 1,
              "column": 9
            },
            "identifierName": "c"
          },
          "name": "c"
        },
        "optional": false
      }
    },
    {
      "type": "ExpressionStatement",
      "start": 11,
      "end": 21,
      "loc": {
        "start": {
          "line": 2,
          "column": 0
        },
        "end": {
          "line": 2,
          "column": 10
        }
      },
      "expression": {
        "type": "MemberExpression",
        "start": 11,
        "end": 20,
        "loc": {
          "start": {
            "line": 2,
            "column": 0
          },
          "end": {
            "line": 2,
            "column": 9
          }
        },
        "object": {
          "type": "TSNonNullExpression",
          "start": 11,
          "end": 18,
          "loc": {
            "start": {
              "line": 2,
              "column": 0
            },
            "end": {
              "line": 2,
              "column": 7
            }
          },
          "expression": {
            "type": "ChainExpression",
            "start": 12,
            "end": 16,
            "loc": {
              "start": {
                "line": 2,
                "column": 1
              },
              "end": {
                "line": 2,
                "column": 5
              }
            },
            "expression": {
              "type": "MemberExpression",
              "start": 12,
              "end": 16,
              "loc": {
                "start": {
                  "line": 2,
                  "column": 1
                },
                "end": {
                  "line": 2,
                  "column": 5
                }
              },
              "object": {
                "type": "Identifier",
                "start": 12,
                "end": 13,
                "loc": {
                  "start": {
                    "line": 2,
                    "column": 1
                  },
                  "end": {
                    "line": 2,
                    "column": 2
                  },
                  "identifierName": "a"
                },
                "name": "a"
              },
              "computed": false,
              "property": {
                "type": "Identifier",
                "start": 15,
                "end": 16,
                "loc": {
                  "start": {
                    "line": 2,
                    "column": 4
                  },
                  "end": {
                    "line": 2,
                    "column": 5
                  },
                  "identifierName": "b"
                },
                "name": "b"
              },
              "optional": true
            },
            "extra": {
              "parenthesized": true,
              "parenStart": 11
            }
          }
        },
        "computed": false,
        "property": {
          "type": "Identifier",
          "start": 19,
          "end": 20,
          "loc": {
            "start": {
              "line": 2,
              "column": 8
            },
            "end": {
              "line": 2,
              "column": 9
            },
            "identifierName": "c"
          },
          "name": "c"
        },
        "optional": false
      }
    }
  ]
}
```

</details>

<details>
<summary>`@typescript-eslint/typescript-estree` @ `4.0.1`</summary>

```json
{
  "type": "Program",
  "body": [
    {
      "type": "ExpressionStatement",
      "expression": {
        "type": "MemberExpression",
        "object": {
          "type": "ChainExpression",
          "expression": {
            "type": "TSNonNullExpression",
            "expression": {
              "type": "MemberExpression",
              "object": {
                "type": "Identifier",
                "name": "a",
                "range": [
                  1,
                  2
                ]
              },
              "property": {
                "type": "Identifier",
                "name": "b",
                "range": [
                  4,
                  5
                ]
              },
              "computed": false,
              "optional": true,
              "range": [
                1,
                5
              ]
            },
            "range": [
              1,
              6
            ]
          },
          "range": [
            1,
            6
          ]
        },
        "property": {
          "type": "Identifier",
          "name": "c",
          "range": [
            8,
            9
          ]
        },
        "computed": false,
        "optional": false,
        "range": [
          0,
          9
        ]
      },
      "range": [
        0,
        10
      ]
    },
    {
      "type": "ExpressionStatement",
      "expression": {
        "type": "MemberExpression",
        "object": {
          "type": "TSNonNullExpression",
          "expression": {
            "type": "ChainExpression",
            "expression": {
              "type": "MemberExpression",
              "object": {
                "type": "Identifier",
                "name": "a",
                "range": [
                  12,
                  13
                ]
              },
              "property": {
                "type": "Identifier",
                "name": "b",
                "range": [
                  15,
                  16
                ]
              },
              "computed": false,
              "optional": true,
              "range": [
                12,
                16
              ]
            },
            "range": [
              12,
              16
            ]
          },
          "range": [
            11,
            18
          ]
        },
        "property": {
          "type": "Identifier",
          "name": "c",
          "range": [
            19,
            20
          ]
        },
        "computed": false,
        "optional": false,
        "range": [
          11,
          20
        ]
      },
      "range": [
        11,
        21
      ]
    }
  ],
  "sourceType": "script",
  "range": [
    0,
    21
  ]
}
```

</details>

Reply:
Thank you for the information, I'll take a deep look